parameterised modules in the QL language reference (#10088)
github/codeql: docs/codeql/ql-language-reference/signatures.rst

:tocdepth: 1

.. index:: signature

.. _signatures:

Signatures
##########

Parameterised modules use signatures as a type system for their parameters.
There are three categories of signatures: **predicate signatures**, **type signatures**, and **module signatures**.

Predicate signatures
====================

> Predicate signatures declare module parameters that are to be substituted with predicates at module instantiation.

Reviewer: Simpler as:

```suggestion
Predicate signatures declare module parameters that will be substituted with predicates when the module is instantiated.
```

> The substitution of predicate signatures relies on structural typing, i.e. predicates do not have to be explicitly
> defined as implementing a predicate signature - they just have to match the return and argument types.

Reviewer: ```suggestion
The substitution of predicate signatures relies on structural typing. That is, predicates do not have to be explicitly
defined as implementing a predicate signature - they just have to match the return and argument types.
```

Predicate signatures are defined much like predicates themselves, but they do not have a body.
In detail, a predicate signature definition consists of:

#. The keyword ``signature``.
#. The keyword ``predicate`` (allows subsitution with a :ref:`predicate without result <predicates-without-result>`),
   or the type of the result (allows subsitution with a :ref:`predicate with result <predicates-with-result>`).
#. The name of the predicate signature. This is an `identifier <https://codeql.github.com/docs/ql-language-reference/ql-language-specification/#identifiers>`_
   starting with a lowercase letter.
#. The arguments to the predicate signature, if any, separated by commas.
   For each argument, specify the argument type and an identifier for the argument variable.
#. A semicolon ``;``.

For example:

.. code-block:: ql

    signature int operator(int lhs, int rhs);

Type signatures
===============

> Type signatures declare module parameters that are to be substituted with types at module instantiation.

Reviewer: ```suggestion
Type signatures declare module parameters that will be substituted with types when the module is instantiated.
```

> Type signature are the simplest category of signatures, as the only thing they allow is the specification of supertypes.

Reviewer: ```suggestion
Type signatures are used to specify supertypes and are the simplest category of signatures.
```

> The substitution of type signatures relies on structural typing, i.e. types do not have to be explicitly defined as

Reviewer: ```suggestion
The substitution of type signatures relies on structural typing. That is, types do not have to be explicitly defined as
```

implementing a type signature - they just need to have the specified (transitive) supertypes.

In detail, a type signature definition consists of:

#. The keyword ``signature``.
#. The keyword ``class``.
#. The name of the type signature. This is an `identifier <https://codeql.github.com/docs/ql-language-reference/ql-language-specification/#identifiers>`_
   starting with a uppercase letter.
#. Optionally, the keyword ``extends`` followed by a list of types, separated by commas.
#. A semicolon ``;``.

For example:

.. code-block:: ql

    signature class ExtendsInt extends int;

Module signatures
=================

> Module signatures declare module parameters that are to be substituted with modules at module instantiation.

Reviewer: ```suggestion
Module signatures declare module parameters that will be substituted with modules when the module is instantiated.
```

Module signatures specify a collection of types and predicates that a module needs to contain under given names and
matching given signatures.

> Contrary to type signatures and predicte signatures, the substitution of type signatures relies on nominal typing,

Reviewer: ```suggestion
Unlike type signatures and predicate signatures, the substitution of type signatures relies on nominal typing.
```

> i.e. modules need to declare at their definition the module signatures they implement.

Reviewer: ```suggestion
That is, the definition of a module must declare the module signatures it implements.
```
